API/Services/KoreaderService.cs

```diff
@@ -0,0 +1,52 @@
+using System;
+using System.Threading.Tasks;
+using API.Data;
+using API.DTOs.Koreader;
+using API.Helpers;
+
+namespace API.Services;
+
+#nullable enable
+
+public interface IKoreaderService
+{
+    Task SaveProgress(KoreaderBookDto koreaderBookDto, int userId);
+    Task<KoreaderBookDto> GetProgress(string bookHash, int userId);
+}
+
+public class KoreaderService : IKoreaderService
+{
+    private IReaderService _readerService;
+    private IUnitOfWork _unitOfWork;
+
+    public KoreaderService(IReaderService readerService, IUnitOfWork unitOfWork)
+    {
+        _readerService = readerService;
+        _unitOfWork = unitOfWork;
+    }
+
+    public async Task SaveProgress(KoreaderBookDto koreaderBookDto, int userId)
+    {
+        var file = await _unitOfWork.MangaFileRepository.GetByHash(koreaderBookDto.Document);
+        var userProgressDto = await _unitOfWork.AppUserProgressRepository.GetUserProgressDtoAsync(file.ChapterId, userId);
+        KoreaderHelper.UpdateProgressDto(koreaderBookDto.Progress, userProgressDto);
+        _readerService.SaveReadingProgress(userProgressDto, userId);
+        await _unitOfWork.CommitAsync();
+    }
+
+    public async Task<KoreaderBookDto> GetProgress(string bookHash, int userId)
+    {
+        var file = await _unitOfWork.MangaFileRepository.GetByHash(bookHash);
+        var progressDto = await _unitOfWork.AppUserProgressRepository.GetUserProgressDtoAsync(file.ChapterId, userId);
+        var koreaderProgress = KoreaderHelper.GetKoreaderPosition(progressDto);
+        return new KoreaderBookDto
+        {
+            Document = bookHash,
+            Device_id = "kavita",
+            Device = "kavita",
```

Comment:
Kavita should be capital. Is device_id supposed to be a unique identifier? Could we use the ServerId? That is unique unless someone changes their token in appsettings.json (which usually only happens on reinstalling incorrectly).

Reply:
ServerId would make more sense if it is a unique identifier. Could you tell me where to get that from? I can't find anything by searching `ServerId` or its synonyms.

Reply:
You should find this on BuildInfo I believe, but maybe getting from the db is better. You can get from setting item. I believe it also goes as InstallId.